src/text_shaper/directwrite_shaper.cpp

```diff
@@ -66,6 +66,60 @@ namespace
         }
         return font_slant::normal;
     }
+
+    void renderGlyphRunToBitmap(IDWriteGlyphRunAnalysis* _glyphAnalysis, RECT& _textureBounds, bitmap_format _targetFormat, std::vector<uint8_t>::iterator& _it)
+    {
+        const auto width = _textureBounds.right - _textureBounds.left;
+        const auto height = _textureBounds.bottom - _textureBounds.top;
+
+        std::vector<uint8_t> tmp;
+        tmp.resize(height *width * 3);
+
+        auto hr = _glyphAnalysis->CreateAlphaTexture(
+            DWRITE_TEXTURE_CLEARTYPE_3x1,
+            &_textureBounds,
+            tmp.data(),
+            tmp.size()
+        );
+
+        for (auto i = 0; i < height; i++)
+            for (auto j = 0; j < width; j++)
+            {
+                const auto base = ((height - 1 - i) *width + j) * 3;
+                const auto r = tmp[base];
+                const auto g = tmp[base + 1];
+                const auto b = tmp[base + 2];
+                if (r > 0)
+                {
+                    *_it++ = r;
+                }
+                else
+                {
+                    _it++;
+                }
+                if (g > 0)
+                {
+                    *_it++ = g;
+                }
+                else
+                {
+                    _it++;
+                }
+                if (b > 0)
+                {
+                    *_it++ = b;
+                }
+                else
+                {
+                    _it++;
+                }
+
+                if (_targetFormat == bitmap_format::rgba)
```

Comment:
Seems that I have to create a `rgba` format bitmap for colorAtlas.

Reply:
Yes. Colored emoji are placed into the scene as RGBA texture. In Freetype i have to convert from BGRA to RGBA too (i think OpenGL ES had Limited pixel format support). On Linux OpenGL ES (core Profile) is used by default.

Reply:
BTW i just realize, that is a perfect opportunity for  providing a SIMD implementation for that task (you don't have to do that,l. I am just writing that one down here so it won't be forgotten).